File: app/src/main/java/org/scp/gymlog/ui/top/TopActivity.java
package org.scp.gymlog.ui.top;

import static org.scp.gymlog.util.Constants.ONE_HUNDRED;

import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.scp.gymlog.R;
import org.scp.gymlog.exceptions.InternalException;
import org.scp.gymlog.exceptions.LoadException;
import org.scp.gymlog.model.Bit;
import org.scp.gymlog.model.Exercise;
import org.scp.gymlog.model.Variation;
import org.scp.gymlog.room.AppDatabase;
import org.scp.gymlog.room.DBThread;
import org.scp.gymlog.ui.common.DBAppCompatActivity;
import org.scp.gymlog.ui.common.dialogs.EditExercisesLastsDialogFragment;
import org.scp.gymlog.ui.top.rows.ITopRow;
import org.scp.gymlog.ui.top.rows.TopBitRow;
import org.scp.gymlog.ui.top.rows.TopEmptySpaceRow;
import org.scp.gymlog.ui.top.rows.TopHeaderRow;
import org.scp.gymlog.ui.top.rows.TopVariationRow;
import org.scp.gymlog.ui.training.TrainingActivity;
import org.scp.gymlog.util.Constants.IntentReference;
import org.scp.gymlog.util.Data;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;

public class TopActivity extends DBAppCompatActivity {
    private Exercise exercise;
    private final List<ITopRow> listData = new ArrayList<>();
    private boolean internationalSystem;
    private TopRecyclerViewAdapter adapter;

    @Override
    protected int onLoad(Bundle savedInstanceState, AppDatabase db) {
        int exerciseId = getIntent().getExtras().getInt("exerciseId");
        exercise = Data.getInstance().getExercises().stream()
                .filter(ex -> ex.getId() == exerciseId)
                .findFirst()
                .orElseThrow(() -> new InternalException("Exercise id not found"));

        transformBitsToRows(getBits(db, exerciseId));
        return CONTINUE;
    }

    protected Stream<Bit> getBits(AppDatabase db, int exerciseId) {
        return db.bitDao().findTops(exerciseId).stream()
                .map(bit -> new Bit().fromEntity(bit));
    }

    protected Comparator<? super Bit> order() {
        return Comparator.comparing(bit -> bit.getWeight().getValue().negate());
    }

    private void transformBitsToRows(Stream<Bit> bits) {
        final Set<Integer> variations = new HashSet<>();
        listData.clear();
        bits.sorted(Comparator.comparingInt(Bit::getVariationId).thenComparing(order()))
                .forEachOrdered(bit -> {
                    int variationId = bit.getVariationId();
                    if (!variations.contains(variationId)) {
                        variations.add(variationId);
                        if (variationId != 0) {
                            Variation variation = Data.getVariation(exercise, variationId);
                            listData.add(new TopVariationRow(variation));
                        }
                        listData.add(new TopHeaderRow());
                    }
                    listData.add(new TopBitRow(bit));
                });
        listData.add(new TopEmptySpaceRow());
    }

    @Override
    protected void onDelayedCreate(Bundle savedInstanceState) {
        setContentView(R.layout.activity_tops);
        setTitle(R.string.title_top_records);

        SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
        internationalSystem = preferences.getBoolean("internationalSystem", true);

        setHeaderInfo();

        final RecyclerView historyRecyclerView = findViewById(R.id.variantTopList);
        historyRecyclerView.setLayoutManager(new LinearLayoutManager(this));

        adapter = new TopRecyclerViewAdapter(listData, exercise, internationalSystem);
        historyRecyclerView.setAdapter(adapter);

        adapter.setOnClickListener(this::onElementClicked);
        adapter.setOnLongClickListener(this::onElementLongClicked);
    }

    protected void onElementClicked(Bit topBit) {
        Intent intent = new Intent(this, TrainingActivity.class);
        intent.putExtra("trainingId", topBit.getTrainingId());
        intent.putExtra("focusBit", topBit.getId());
        startActivityForResult(intent, IntentReference.TRAINING);
    }

    protected void onElementLongClicked(Bit topBit) {
        Intent intent = new Intent(this, TopSpecificActivity.class);
        intent.putExtra("exerciseId", topBit.getExerciseId());
        intent.putExtra("weight", topBit.getWeight().getValue().multiply(ONE_HUNDRED).intValue());
        intent.putExtra("variationId", topBit.getVariationId());
        startActivityForResult(intent, IntentReference.TOP_RECORDS);
    }

    public void onActivityResult(IntentReference intentReference, Intent data) {
        if (data.getBooleanExtra("refresh", false)) {
            if (intentReference == IntentReference.TOP_RECORDS) {
                adapter.notifyItemRangeChanged(0, listData.size());

            } else if (intentReference == IntentReference.TRAINING) {
                DBThread.run(this, db -> {
                    transformBitsToRows(getBits(db, exercise.getId()));
                    runOnUiThread(() -> adapter.notifyDataSetChanged());
                });
            }
        }
    }

    private void setHeaderInfo() {
        View fragment = findViewById(R.id.fragmentExercise);
        ImageView image = findViewById(R.id.image);
        TextView time = findViewById(R.id.time);
        TextView title = findViewById(R.id.content);


        title.setText(exercise.getName());
        time.setVisibility(View.GONE);
        String fileName = "previews/" + exercise.getImage() + ".png";
        try {
            InputStream ims = getAssets().open(fileName);
            Drawable d = Drawable.createFromStream(ims, null);
            image.setImageDrawable(d);

        } catch (IOException e) {
            throw new LoadException("Could not read \""+fileName+"\"", e);
        }

        fragment.setOnClickListener(v -> {
            EditExercisesLastsDialogFragment dialog = new EditExercisesLastsDialogFragment(R.string.title_exercises,
                    val -> {
                        if (val != null) {
                            Intent data = new Intent();
                            data.putExtra("refresh", true);
                            setResult(RESULT_OK, data);
                            runOnUiThread(() -> adapter.notifyItemRangeChanged(0, listData.size()));
                        }
                    },
                    ()->{},
                    exercise, internationalSystem);

            dialog.show(getSupportFragmentManager(), null);
        });
    }

}

File: app/src/main/java/org/scp/gymlog/ui/top/TopActivity.kt
package org.scp.gymlog.ui.top

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.preference.PreferenceManager
import androidx.viewbinding.ViewBinding
import org.scp.gymlog.R
import org.scp.gymlog.databinding.ListitemTopBitBinding
import org.scp.gymlog.databinding.ListitemTopHeadersBinding
import org.scp.gymlog.databinding.ListitemTopSpaceBinding
import org.scp.gymlog.databinding.ListitemTopVariationBinding
import org.scp.gymlog.exceptions.InternalException
import org.scp.gymlog.exceptions.LoadException
import org.scp.gymlog.model.Bit
import org.scp.gymlog.model.Exercise
import org.scp.gymlog.room.AppDatabase
import org.scp.gymlog.ui.common.DBAppCompatActivity
import org.scp.gymlog.ui.common.components.listView.MultipleListHandler
import org.scp.gymlog.ui.common.components.listView.MultipleListView
import org.scp.gymlog.ui.common.components.listView.SimpleListView
import org.scp.gymlog.ui.preferences.PreferencesDefinition
import org.scp.gymlog.ui.top.rows.*
import org.scp.gymlog.ui.training.TrainingActivity
import org.scp.gymlog.util.Constants
import org.scp.gymlog.util.Constants.IntentReference
import org.scp.gymlog.util.Constants.TODAY
import org.scp.gymlog.util.Data
import org.scp.gymlog.util.DateUtils
import org.scp.gymlog.util.DateUtils.getDateString
import org.scp.gymlog.util.DateUtils.getLetterFrom
import org.scp.gymlog.util.FormatUtils.bigDecimal
import org.scp.gymlog.util.FormatUtils.integer
import org.scp.gymlog.util.WeightUtils.calculate
import org.scp.gymlog.util.extensions.DatabaseExts.dbThread
import org.scp.gymlog.util.extensions.PreferencesExts.loadBoolean
import java.io.IOException

open class TopActivity : DBAppCompatActivity() {

    private lateinit var exercise: Exercise
    private val listData: MutableList<ITopRow> = ArrayList()
    private var internationalSystem = false
    private lateinit var topListView: MultipleListView<ITopRow>

    override fun onLoad(savedInstanceState: Bundle?, db: AppDatabase): Int {
        val exerciseId = intent.extras!!.getInt("exerciseId")
        exercise = Data.exercises
            .filter { exercise -> exercise.id == exerciseId }
            .getOrElse(0) { throw InternalException("Exercise id not found") }

        transformBitsToRows(getBits(db, exerciseId))
        return CONTINUE
    }

    override fun onDelayedCreate(savedInstanceState: Bundle?) {
        setContentView(R.layout.activity_tops)
        setTitle(R.string.title_top_records)

        internationalSystem = loadBoolean(PreferencesDefinition.UNIT_INTERNATIONAL_SYSTEM)

        setHeaderInfo()

        topListView = findViewById(R.id.variantTopList)
        topListView.init(listData, TopListHandler())
    }

    protected open fun getBits(db: AppDatabase, exerciseId: Int): List<Bit> {
        return db.bitDao().findTops(Data.currentGym, exerciseId)
            .map { bitEntity -> Bit(bitEntity) }
    }

    protected open fun order(): Comparator<in Bit> {
        return Comparator.comparing { bit -> bit.weight.getValue(internationalSystem).negate() }
    }

    private fun transformBitsToRows(bits: List<Bit>) {
        val variations: MutableSet<Int> = HashSet()
        listData.clear()
        bits.sortedWith (order())
            .sortedWith(compareBy { it.variation.id })
            .sortedWith(compareBy { !it.variation.default })
            .forEach { bit ->
                val variationId = bit.variation.id
                if (!variations.contains(variationId)) {
                    variations.add(variationId)
                    if (!bit.variation.default) {
                        val variation = Data.getVariation(exercise, variationId)
                        listData.add(TopVariationRow(variation))
                    }
                    listData.add(TopHeaderRow())
                }
                listData.add(TopBitRow(bit))
            }
        listData.add(TopEmptySpaceRow())
    }

    protected fun onElementClicked(topBit: Bit) {
        val intent = Intent(this, TrainingActivity::class.java)
        intent.putExtra("trainingId", topBit.trainingId)
        intent.putExtra("focusBit", topBit.id)
        startActivityForResult(intent, IntentReference.TRAINING)
    }

    protected open fun onElementLongClicked(topBit: Bit) {
        val intent = Intent(this, TopSpecificActivity::class.java)
        intent.putExtra("exerciseId", topBit.variation.exercise.id)
        intent.putExtra("weight", topBit.weight.value.multiply(Constants.ONE_HUNDRED).toInt())
        intent.putExtra("variationId", topBit.variation.id)
        startActivityForResult(intent, IntentReference.TOP_RECORDS)
    }

    override fun onActivityResult(intentReference: IntentReference, data: Intent) {
        if (data.getBooleanExtra("refresh", false)) {
            if (intentReference === IntentReference.TOP_RECORDS ||
                intentReference === IntentReference.TRAINING) {
                dbThread { db ->
                    transformBitsToRows(getBits(db, exercise.id))
                    topListView.setListData(listData)
                    runOnUiThread { topListView.notifyDataSetChanged() }
                }
            }
        }
    }

    private fun setHeaderInfo() {
        val image = findViewById<ImageView>(R.id.image)
        val time = findViewById<TextView>(R.id.time)
        val title = findViewById<TextView>(R.id.exerciseName)

        findViewById<TextView>(R.id.variationName).visibility = View.GONE

        title.text = exercise.name
        time.visibility = View.GONE
        val fileName = "previews/" + exercise.image + ".png"
        try {
            val ims = assets.open(fileName)
            val d = Drawable.createFromStream(ims, null)
            image.setImageDrawable(d)
        } catch (e: IOException) {
            throw LoadException("Could not read \"$fileName\"", e)
        }
    }

    inner class TopListHandler : MultipleListHandler<ITopRow> {
        override val useListState = false

        override fun generateListItemInflater(item: ITopRow): (LayoutInflater, ViewGroup?, Boolean) -> ViewBinding {
            return when(item.type) {
                ITopRow.Type.BIT -> ListitemTopBitBinding::inflate
                ITopRow.Type.VARIATION -> ListitemTopVariationBinding::inflate
                ITopRow.Type.HEADER -> ListitemTopHeadersBinding::inflate
                else -> ListitemTopSpaceBinding::inflate
            }
        }

        override fun buildListView(
            binding: ViewBinding,
            item: ITopRow,
            index: Int,
            state: SimpleListView.ListElementState?
        ) {
            when (item.type) {
                ITopRow.Type.VARIATION -> {
                    binding as ListitemTopVariationBinding
                    item as TopVariationRow
                    binding.variationName.text = item.variation.name
                }
                ITopRow.Type.BIT -> {
                    binding as ListitemTopBitBinding

                    item as TopBitRow
                    val bit = item.bit

                    val weight = bit.weight.calculate(
                        bit.variation.weightSpec,
                        bit.variation.bar)

                    binding.weight.bigDecimal = weight.getValue(internationalSystem)
                    binding.reps.integer = bit.reps

                    @SuppressLint("SetTextI18n")
                    binding.time.text = bit.timestamp.getDateString() + " (" +
                        TODAY.getLetterFrom(bit.timestamp) + ")"

                    binding.note.text = bit.note

                    binding.root.setOnClickListener {
                        onElementClicked(item.bit)
                    }

                    binding.root.setOnLongClickListener {
                        onElementLongClicked(item.bit)
                        true
                    }
                }
                else -> {}
            }

        }
    }
}